dm/syncer/safe_mode.go

```diff
@@ -43,35 +57,77 @@ func (s *Syncer) enableSafeModeInitializationPhase(tctx *tcontext.Context) {
+	} else {
+		duration, err = time.ParseDuration(initPhaseSeconds)
+		if err != nil {
+			s.tctx.L().Error("enable safe-mode failed due to duration parse failed", zap.String("duration", initPhaseSeconds))
+			return
+		}
+	}
+	exitPoint := s.checkpoint.SafeModeExitPoint()
+	beginLocation := s.checkpoint.GlobalPoint()
+	failpoint.Inject("SafeModeDurationSetBeginLoc", func() {
+		beginLocation = binlog.MustZeroLocation(mysql.MySQLFlavor)
+	})
+	if exitPoint != nil {
+		s.tctx.L().Info("compare exitPoint and beginLocation", zap.Stringer("exit point", exitPoint), zap.Stringer("begin location", beginLocation))
+		if binlog.CompareLocation(*exitPoint, beginLocation, s.cfg.EnableGTID) == 0 {
+			s.tctx.L().Info("exitPoint equal to beginLocation, so disable the safe mode")
+			s.checkpoint.SaveSafeModeExitPoint(nil)
+			return
+		}
+
 		//nolint:errcheck
 		s.safeMode.Add(tctx, 1) // enable and will revert after pass SafeModeExitLoc
-		s.tctx.L().Info("enable safe-mode for safe mode exit point, will exit at", zap.Stringer("location", *s.checkpoint.SafeModeExitPoint()))
+		s.tctx.L().Info("enable safe-mode for safe mode exit point, will exit at", zap.Stringer("location", *exitPoint))
 	} else {
-		//nolint:errcheck
-		s.safeMode.Add(tctx, 1) // enable and will revert after 2 * CheckpointFlushInterval
-		go func() {
-			defer func() {
-				err := s.safeMode.Add(tctx, -1)
-				if err != nil {
-					// send error to the fatal chan to interrupt the process
-					s.runFatalChan <- unit.NewProcessError(err)
-				}
-				if !s.safeMode.Enable() {
-					s.tctx.L().Info("disable safe-mode after task initialization finished")
+		s.tctx.L().Info("enable safe-mode because of task initialization", zap.Duration("duration", duration))
+
+		if int64(duration) > 0 {
+			//nolint:errcheck
+			s.safeMode.Add(tctx, 1) // enable and will revert after 2 * CheckpointFlushInterval
+			go func() {
+				defer func() {
+					err = s.safeMode.Add(tctx, -1)
+					if err == nil && !s.safeMode.Enable() {
+						s.tctx.L().Info("disable safe-mode after task initialization finished")
+					}
+				}()
+
+				select {
+				case <-tctx.Context().Done():
+				case <-time.After(duration):
 				}
 			}()
-			initPhaseSeconds := s.cfg.CheckpointFlushInterval * 2
-			failpoint.Inject("SafeModeInitPhaseSeconds", func(val failpoint.Value) {
-				seconds, _ := val.(int)
-				initPhaseSeconds = seconds
-				s.tctx.L().Info("set initPhaseSeconds", zap.String("failpoint", "SafeModeInitPhaseSeconds"), zap.Int("value", seconds))
-			})
-			s.tctx.L().Info("enable safe-mode because of task initialization", zap.Int("duration in seconds", initPhaseSeconds))
-			select {
-			case <-tctx.Context().Done():
-			case <-time.After(time.Duration(initPhaseSeconds) * time.Second):
+		}
+	}
+
+	if int64(duration) == 0 {
+		failpoint.Inject("SafeModeInitPhaseSeconds", func(val failpoint.Value) {
+			seconds := val.(string)
+			if seconds == "0s" {
+				failpoint.Return()
 			}
-		}()
+		})
```

Comment:
make sure safe mode disable when failpoint `SafeModeInitPhaseSeconds = 0s`

Because `SafeModeInitPhaseSeconds = 0s` was used in so many places

Reply:
"make sure safe mode disable"  is inaccurate.

maybe disable InitPhaseSeconds affect

Reply:
> > make sure safe mode disable when failpoint `SafeModeInitPhaseSeconds = 0s`
> > Because `SafeModeInitPhaseSeconds = 0s` was used in so many places
> 
> I can't understand why failpoint.Return() is related to "make sure safe mode disable"

Because now, `safe-mode-duration=0s` will report an error when `task not fresh`

Reply:
What's the expected test scenario for it?

Reply:
removed it https://github.com/pingcap/tiflow/pull/6225/commits/f70bfb5e4b7576e6df6f2e211f569510225778fe @lance6716